src/validation.cpp

```diff
@@ -3690,6 +3693,124 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+    }
+    return true;
+}
+
+/** Unapply the effects of a block on the utxo cache, ignoring that it may already have been applied. */
+static bool RollbackBlock(const CBlockIndex* pindex, CCoinsViewCache& inputs, const CChainParams& params)
+{
+    // TODO: Merge with DisconnectBlock
+    CBlock block;
+    if (!ReadBlockFromDisk(block, pindex, params.GetConsensus())) {
+        return error("RollbackBlock(): ReadBlockFromDisk failed at %d, hash=%s", pindex->nHeight, pindex->GetBlockHash().ToString());
+    }
+
+    CBlockUndo blockUndo;
+    CDiskBlockPos pos = pindex->GetUndoPos();
+    if (pos.IsNull())
+        return error("RollbackBlock(): no undo data available");
+    if (!UndoReadFromDisk(blockUndo, pos, pindex->pprev->GetBlockHash()))
+        return error("RollbackBlock(): failure reading undo data");
+
+    if (blockUndo.vtxundo.size() + 1 != block.vtx.size())
+        return error("RollbackBlock(): block and undo data inconsistent");
+
+    for (size_t i = 0; i < block.vtx.size(); ++i) {
+        const CTransaction& tx = *block.vtx[i];
+        inputs.ModifyCoins(tx.GetHash())->Clear();
+        if (i != 0) {
+            CTxUndo &txundo = blockUndo.vtxundo[i - 1];
+            if (txundo.vprevout.size() != tx.vin.size()) {
+                return error("RollbackBlock(): block and undo data inconsistent");
+            }
+            for (size_t n = 0; n < txundo.vprevout.size(); ++n) {
+                ApplyTxInUndo(txundo.vprevout[n], inputs, COutPoint(tx.GetHash(), n)); // Ignore return value
+            }
+        }
+    }
+    return true;
+}
+
+
+bool ReplayBlocks(const CChainParams& params, CCoinsView* view)
+{
+    LOCK(cs_main);
+
+    CCoinsViewCache cache(view);
+
+    uint256 hashBest = view->GetBestBlock();
+    if (!hashBest.IsNull()) return true;
+    std::vector<uint256> hashHeads = view->GetHeadBlocks();
+    if (hashHeads.empty()) return true;
+
+    uiInterface.ShowProgress(_("Replaying blocks..."), 0);
+    LogPrintf("Replaying blocks\n");
+
+    std::vector<const CBlockIndex*> pindexHeads;
+    const CBlockIndex* pindexFork = nullptr;
+    bool fGenesis = false;
+
+    // Find last common ancestor of all heads.
+    for (const uint256& hash : hashHeads) {
+        if (hash.IsNull()) {
+            fGenesis = true;
+            continue;
+        }
+        auto it = mapBlockIndex.find(hash);
+        if (it == mapBlockIndex.end()) {
+            return error("ReplayBlocks(): chainstate boundaries not in block index");
+        }
+        pindexHeads.push_back(it->second);
+        pindexFork = pindexFork ? LastCommonAncestor(pindexFork, it->second) : it->second;
+    }
+
+    // Build a set of all blocks to rollback.
+    std::set<const CBlockIndex*, CBlockIndexWorkComparator> vpindexRollback;
+    for (size_t i = 1; i < pindexHeads.size(); ++i) {
+        const CBlockIndex *pindexHead = pindexHeads[i];
+        while (fGenesis ? pindexHead != nullptr : pindexHead != pindexFork) {
```

Comment:
No. Let's say you have a chain A->B->C that was being flushed (meaning the old tip was A, you crashed in the middle of writing the changes for B and C, with C the intended new tip). In this case, at recovery time, GetBlockHeads() will return `[C,A]`. `pindexFork` will be A. The loop above will only process A, but because A is already the fork point, nothing is added to the disconnect set.